File: service/src/main/java/org/pulcifer/climate/controller/ClimateController.java
package org.pulcifer.climate.controller;

import org.pulcifer.climate.dto.SimilarCity;
import org.pulcifer.climate.model.City;
import org.pulcifer.climate.svc.CityService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
public class ClimateController {

    @Autowired
    CityService svc;

    @GetMapping("/city/{id}")
    public City city(@PathVariable Integer id) {
        return svc.getCityById(id).get();
    }

    @GetMapping("/city")
    public City cityByName(@RequestParam(value = "name") String name) {
        return svc.getCityByName(name).get();
    }

    @GetMapping("/citiesContaining")
    public Map<Integer, String> citiesContaining(@RequestParam(value = "name") String name) {
       return svc.citiesContaining(name).get().stream()
                .collect(Collectors.toMap(City::getCityId, City::getCityName));
    }

    @GetMapping("/similar")
    public List<SimilarCity> similar(@RequestParam(value = "cityId")Integer cityId,
                          @RequestParam(value = "outliers")Integer outliers) {
        return svc.retrieveSimilarCities(cityId, outliers);
    }
}

File: src/main/kotlin/org/pulcifer/climate/controller/ClimateController.kt
package org.pulcifer.climate.controller

import org.pulcifer.climate.model.City
import org.pulcifer.climate.model.SimilarCity
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.RestController
import org.pulcifer.climate.svc.CityService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestParam

@RestController
class ClimateController {
    @Autowired
    var svc: CityService? = null
    @GetMapping("/city/{id}")
    fun city(@PathVariable id: Int?): City? {
        return svc!!.getCityById(id)
    }

    @GetMapping("/city")
    fun cityByName(@RequestParam(value = "name") name: String?): List<City?> {
        return svc!!.citiesContaining(name)
    }

//    @GetMapping("/citiesContaining")
//    fun citiesContaining(@RequestParam(value = "name") name: String?): Map<Int?, String> {
//        return svc!!.citiesContaining(name).stream()
//            .collect(Collectors.toMap(
//                Function { obj: City? -> obj.getCityId() }, Function { city: City? -> formatCityCountry(city) })
//            )
//    }

    private fun formatCityCountry(city: City?): String {
        return String.format("%s/%s", city?.cityName, city?.country)
    }

    @GetMapping("/similar")
    fun similar(
        @RequestParam(value = "cityId") cityId: Int?,
        @RequestParam(value = "outliers") outliers: Int
    ): List<SimilarCity?>? {
        return svc!!.retrieveSimilarCities(cityId, outliers)
    }

    @GetMapping("/citiesCompare")
    fun compareCities(@RequestParam(value = "cityId") cityId: Int, @RequestParam(value = "cityIds") cityIds: List<Int>): List<SimilarCity?>? {
        return svc!!.compareCities(cityId, cityIds)
    }
}